### PR TITLE
[Refactor] Fix clippy::too_many_arguments and rework error page rendering

### LINKED ISSUE #588: Error page isn't rendered with the correct theme
It seems the error page is always rendered with the default theme and not the user's selected theme.

### PR DESCRIPTION
**Pull request by aliemjay**
Too many arguments are moved around and many of them are already stored in `MiniserveConfig`. Many of these are used to render error pages.

To fix this issue, it was necessary to rework error page rendering:

1. Implement `ResponseError` for `ContextualError` so that it can be
   returned from service handlers as is and will then be automatically
   logged to the console and converted into an error response.

2. At service handler level, all error responses are now rendered as
   plain text.

3. `error_page_middleware` is now responsible for the rendering of the
   final error page from plain text responses.

Based on #513 

Closes #588

### CODE REVIEW

**Comment by aliemjay** [comment]
Hi @svenstaro,

I currently base my new PRs on some previous PRs to avoid huge merge conflicts when merging. What do you think of such workflow?  I found this necessary because PRs are piling up.

**Comment by aliemjay** [comment]
Is such code refactoring welcome? I hope it is not invasive!

**Comment by aliemjay** [comment]
@svenstaro I hope this gets reviewed next

**Comment by svenstaro** [comment]
Sure but let's get #500 in first. We're almost there! \o/

**Review by svenstaro — CHANGES_REQUESTED** [review]
I get a plain (unthemed) error page for invalid or missing auth here. Granted, this also was the previous behavior but how about we fix this in the same PR? The 404 error, for instance, is already themed. It would make sense for 401/403 to also be themed and it would be more consistent, too.

**Comment by svenstaro** [comment]
While testing this, I noticed #588 which you could address in this PR as well (or it could just be a follow-up).

**Comment by aliemjay** [comment]
> I get a plain (unthemed) error page for invalid or missing auth here. Granted, this also was the previous behavior but how about we fix this in the same PR? The 404 error, for instance, is already themed. It would make sense for 401/403 to also be themed and it would be more consistent, too.

Is it even possible to get themed in invalid auth? I mean unless we embed css.

**Comment by aliemjay** [comment]
> While testing this, I noticed #588 which you could address in this PR as well (or it could just be a follow-up).

This should be easy. I'll implement it here.

**Comment by svenstaro** [comment]
> > I get a plain (unthemed) error page for invalid or missing auth here. Granted, this also was the previous behavior but how about we fix this in the same PR? The 404 error, for instance, is already themed. It would make sense for 401/403 to also be themed and it would be more consistent, too.
> 
> Is it even possible to get themed in invalid auth? I mean unless we embed css.

I suppose the internal route to the media resources doesn't really need to be protected, no secrets will ever be kept there. The theme information is stored inside the browser itself so that should also be always available.

**Comment by svenstaro** [comment]
The rest of it looks like a vast improvement to me! Probably there are some parts which could be golfed a little (like the middleware bit) but I wouldn't immediately know how to best golf it.

File: src/renderer.rs [hunk 1]
```diff
@@ -469,48 +462,34 @@ fn humanize_systemtime(time: Option<SystemTime>) -> Option<String> {
...
+                .(format!("default_theme_dark_{}", conf.default_color_scheme_dark)) {
 
                 div.error {
                     p { (error_code.to_string()) }
                     @for error in error_description.lines() {
                         p { (error) }
                     }
-                    @if display_back_link {
+                    // WARN don't expose random route!
+                    @if !conf.random_route.is_some() {
```
Comment (svenstaro) [inline src/renderer.rs:485]: Good catch!

File: src/listing.rs [hunk 2]
```diff
@@ -185,13 +168,18 @@ pub fn directory_listing(
...
+            .clone()
+            .unwrap_or_else(|| req.connection_info().host().into());
 
         let decoded = percent_decode_str(&encoded_dir).decode_utf8_lossy();
 
         let mut res: Vec<Breadcrumb> = Vec::new();
-        let mut link_accumulator =
-            format!("/{}", random_route.map(|r| r + "/").unwrap_or_default());
+        let mut link_accumulator = match &conf.random_route {
+            Some(random_route) => format!("/{}/", random_route),
```
Comment (svenstaro) [inline src/listing.rs:180]: Yeah, while this is more verbose, I think it's much nicer to read.

Comment (aliemjay) [reply src/listing.rs:180]: This was a necessary evil, because otherwise it would be 7 lines long!

**Comment by svenstaro** [comment]
I snuck in #542. Could you rebase on top of that? Sorry!

**Comment by aliemjay** [comment]
#588 and themed 401 page should be fixed

**Comment by svenstaro** [comment]
Very cool!

**Pull request merged**